src/libponyc/expr/call.c

```diff
@@ -194,6 +194,51 @@ static bool apply_default_arg(pass_opt_t* opt, ast_t* param, ast_t** argp)
   return true;
 }
 
+static ast_t* method_receiver_type(ast_t* method);
+
+bool check_auto_recover_newref(ast_t* ast)
+{
+  while (ast != NULL && ast_id(ast) != TK_CALL)
+    ast = ast_child(ast);
+
+  if (ast == NULL)
+    return false;
+
+  AST_GET_CHILDREN(ast, newref, positional, named);
+  if (ast_id(newref) != TK_NEWREF)
+    return false;
+
+  // sometimes newrefs are nested?
+  ast_t* child = ast_child(newref);
+  if (child != NULL && ast_id(child) == TK_NEWREF)
+    newref = child;
+
+  ast_t* receiver_type = method_receiver_type(newref);
+
+  ast_t* arg = ast_child(positional);
+  while (arg != NULL)
+  {
+    ast_t* arg_type = ast_type(arg);
+    if (is_typecheck_error(arg_type))
+      return false;
+
+    ast_t* arg_type_aliased = alias(arg_type);
+    bool ok = safe_to_autorecover(receiver_type, arg_type_aliased, WRITE);
+    ast_free_unattached(arg_type_aliased);
+
+    if (!ok)
+      return false;
+
+    arg = ast_sibling(arg);
+  }
+
+  token_id tcap = ast_id(cap_fetch(receiver_type));
+  if (tcap == TK_REF || tcap == TK_BOX)
+    return true;
```

Comment:
It might be nice to instead eagerly return false if the capability doesn't match, that way you don't have to check the arguments if the return capability is already wrong.

Reply:
Done.